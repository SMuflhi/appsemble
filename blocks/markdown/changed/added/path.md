Add `data-path` attribute to root node of the DOM output.
