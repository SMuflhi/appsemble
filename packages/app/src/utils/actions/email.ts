import { BaseAction } from '@appsemble/sdk';
import { EmailActionDefinition } from '@appsemble/types';
import axios from 'axios';

import { MakeActionParameters } from '../../types';
import { apiUrl, appId } from '../settings';

export function email({
  prefix,
}: MakeActionParameters<EmailActionDefinition>): BaseAction<'email'> {
  return {
    type: 'email',

    async dispatch(data: any) {
      const url = `${apiUrl}/api/apps/${appId}/action/${prefix}`;
      await axios.post(url, data || {});

      return data;
    },
  };
}
