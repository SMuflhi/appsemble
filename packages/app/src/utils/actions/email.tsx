import type { BaseAction } from '@appsemble/sdk';
import type { EmailActionDefinition } from '@appsemble/types';
import axios from 'axios';

import type { MakeActionParameters } from '../../types';
import settings from '../settings';

export default function email({
  prefix,
}: MakeActionParameters<EmailActionDefinition>): BaseAction<'email'> {
  return {
    type: 'email',

    async dispatch(data: any) {
      const url = `${settings.apiUrl}/api/apps/${settings.id}/action/${prefix}`;
      await axios.post(url, data || {});

      return data;
    },
  };
}
