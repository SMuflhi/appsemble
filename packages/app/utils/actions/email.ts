import axios from 'axios';

import { ActionCreator } from '.';
import { apiUrl, appId } from '../settings';

export const email: ActionCreator<'email'> = ({ prefixIndex }) => [
  async (data) => {
    const url = `${apiUrl}/api/apps/${appId}/action/${prefixIndex}`;
    await axios.post(url, data || {});

    return data;
  },
];
