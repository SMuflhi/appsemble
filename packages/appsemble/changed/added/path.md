Add property `path` to `BootstrapParams`
