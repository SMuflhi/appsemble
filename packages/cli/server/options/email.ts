export const email = (): Promise<void> => Promise.resolve();
