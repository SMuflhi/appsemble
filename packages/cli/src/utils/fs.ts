import { Dirent, promises as fs, Stats } from 'fs';
import { join } from 'path';

import { AppsembleError } from '@appsemble/node-utils';
import type { Promisable } from 'type-fest';

interface OpenDirSafeOptions {
  allowMissing?: boolean;
}

/**
 * Read the contents of a directory.
 *
 * @param directory - The path of the directory to open.
 * @param onFile - A callback which will get called for every file. This will be called with the
 * full file path as its first argument and its `Dirent` object as the second argument.
 * @param options - Additional options.
 */
export async function opendirSafe(
  directory: string,
  onFile: (fullpath: string, stat: Dirent) => Promisable<void>,
  { allowMissing }: OpenDirSafeOptions = {},
): Promise<void> {
  let stats: Stats;
  try {
    stats = await fs.stat(directory);
  } catch (err: unknown) {
    if (allowMissing && (err as NodeJS.ErrnoException).code === 'ENOENT') {
      return;
    }
    throw new AppsembleError(`Expected ${directory} to be a directory`);
  }
  if (!stats.isDirectory()) {
    throw new AppsembleError(`Expected ${directory} to be a directory`);
  }
  const dir = await fs.opendir(directory);
  for await (const file of dir) {
    await onFile(join(directory, file.name), file);
  }
}
