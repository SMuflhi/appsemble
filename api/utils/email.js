import fs from 'fs';
import path from 'path';

import frontmatter from 'front-matter';
import { template } from 'lodash';
import nodemailer from 'nodemailer';
import stubTransport from 'nodemailer-stub-transport';
import { markdown } from 'nodemailer-markdown';

function readTemplate(templateName) {
  return fs.readFileSync(path.join(__dirname, `../templates/${templateName}.md`), 'utf8');
}

export function processTemplate(temp, replacements) {
  const { attributes, body } = frontmatter(temp);

  const content = template(body)({ ...attributes, ...replacements });
  return { attributes: { ...attributes, ...replacements }, content };
}

async function getTransport(smtp) {
  let transport;

  if (smtp) {
    transport = nodemailer.createTransport(smtp);
  } else {
    transport = nodemailer.createTransport(stubTransport());
  }

  transport.use('compile', markdown());
  return transport;
}

export async function sendEmail({ to, cc, bcc, subject }, message, smtp) {
  const transport = await getTransport(smtp);
  const { from } = smtp || '';
  const result = await transport.sendMail({
    from,
    to,
    subject,
    cc,
    bcc,
    markdown: message,
  });

  if (process.env.NODE_ENV !== 'production' || !smtp) {
    // Filter out fields that are unique for snapshot testing
    result.response = result.response
      .toString()
      .replace(
        /(Message-ID:\r?\n? <\w{8}-\w{4}-\w{4}-\w{4}-\w{12}@(.+)>$|Date: .+$|----_NmP.+$|boundary="--_NmP.+$)/gm,
        '',
      );

    if (process.env.NODE_ENV !== 'test') {
      // eslint-disable-next-line no-console
      console.log(`Mail not sent:\n${result.response}`);
    }
  }

  transport.close();

  return result;
}

export async function sendWelcomeEmail({ email, name, url }, smtp) {
  const replacements = {
    greeting: name ? `Hello ${name}` : 'Hello',
    url,
  };

  const { attributes, content } = processTemplate(readTemplate('welcome'), replacements);
  const { subject } = attributes;
  const to = name ? `"${name}" <${email}>` : email;

  return sendEmail({ to, subject }, content, smtp);
}

export async function resendVerificationEmail({ email, name, url }, smtp) {
  const replacements = {
    greeting: name ? `Hello ${name}` : 'Hello',
    url,
  };

  const { attributes, content } = processTemplate(readTemplate('resend'), replacements);
  const { subject } = attributes;
  const to = name ? `"${name}" <${email}>` : email;

  return sendEmail({ to, subject }, content, smtp);
}
