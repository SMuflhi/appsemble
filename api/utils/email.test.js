import fs from 'fs';
import path from 'path';

import { sendEmail, sendWelcomeEmail, resendVerificationEmail } from './email';

describe('sendMail', () => {
  it('should convert markdown to text and html', async () => {
    const result = await sendEmail(
      { to: 'test@example.com', from: 'test@example.com', subject: 'Test' },
      '**Bold Text** Regular text',
      null,
    );

    const {
      envelope: { to },
    } = result;
    const converted = result.response.toString();

    expect(to).toEqual(['test@example.com']);
    expect(converted).toMatch('Content-Type: multipart/alternative;');
    expect(converted).toMatch('Content-Type: text/plain');
    expect(converted).toMatch('Content-Type: text/html');
    expect(converted).toMatch('**Bold Text** Regular text');
    expect(converted).toMatch('<p><strong>Bold Text</strong> Regular text</p>');
    expect(converted).toMatchSnapshot();
  });

  it('should insert template variables', async () => {
    const result = await sendWelcomeEmail(
      { email: 'test@example.com', name: 'John Doe', url: 'https://example.com/test' },
      null,
    );

    const converted = result.response.toString();

    expect(converted).toMatch('Hello John Doe');
    expect(converted).toMatch('https://example.com/test');
    expect(converted).not.toMatch('<%= url %>');
  });

  it('should extract template variables', async () => {
    const result = await sendWelcomeEmail(
      { email: 'test@example.com', name: 'John Doe', url: 'https://example.com/test' },
      null,
    );

    const template = fs.readFileSync(path.join(__dirname, '../templates/welcome.md'), 'utf8');
    const [subject] = template.match(/^subject=.+$/m);
    const converted = result.response.toString();

    expect(subject).toBeTruthy();
    expect(converted).toMatch(`Subject: ${subject.split('=')[1]}`);
  });
});

describe('sendWelcomeEmail', () => {
  it('should match its snapshot', async () => {
    const result = await sendWelcomeEmail(
      { email: 'test@example.com', name: 'John Doe', url: 'https://example.com/test' },
      null,
    );

    const converted = result.response.toString();
    expect(converted).toMatchSnapshot();
  });

  it('should combine name and email', async () => {
    const result = await sendWelcomeEmail(
      { email: 'test@example.com', name: 'John Doe', url: 'https://example.com/test' },
      null,
    );

    const converted = result.response.toString();
    expect(converted).toMatch('John Doe <test@example.com>');
  });
});

describe('resendVerificationEmail', () => {
  it('should match its snapshot', async () => {
    const result = await resendVerificationEmail(
      { email: 'test@example.com', name: 'John Doe', url: 'https://example.com/test' },
      null,
    );

    const converted = result.response.toString();
    expect(converted).toMatchSnapshot();
  });

  it('should combine name and email', async () => {
    const result = await resendVerificationEmail(
      { email: 'test@example.com', name: 'John Doe', url: 'https://example.com/test' },
      null,
    );

    const converted = result.response.toString();
    expect(converted).toMatch('John Doe <test@example.com>');
  });
});
