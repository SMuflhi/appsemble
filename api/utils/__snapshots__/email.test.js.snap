// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resendVerificationEmail should match its snapshot 1`] = `
"Content-Type: multipart/alternative;

To: John Doe <test@example.com>
Subject: Confirm account registration


MIME-Version: 1.0


Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable

Hello John Doe,

Someone has requested to resend the verification key for =
your account. =20
Before you can use your account, we need to verify your =
email address.

Please follow the link below to verify your email address: =
=20
https://example.com/test

**Didn=E2=80=99t request this email?** =20
No worries! Your address may have been entered by mistake. If you ignore or=
 delete this email,
nothing further will happen.

Kind regards,

_The Appsemble Team_


Content-Type: text/html; charset=utf-8
Content-Transfer-Encoding: quoted-printable

<p>Hello John Doe,</p>
<p>Someone has requested to resend the verification =
key for your account.<br>Before you can use your account, we need to verify=
 your email address.</p>
<p>Please follow the link below to verify your =
email address:<br><a href=3D\\"https://example.com/test\\">https://example.=
com/test</a></p>
<p><strong>Didn=E2=80=99t request this email?=
</strong><br>No worries! Your address may have been entered by mistake. If =
you ignore or delete this email,
nothing further will happen.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>


"
`;

exports[`sendMail should convert markdown to text and html 1`] = `
"Content-Type: multipart/alternative;

To: test@example.com
Subject: Test


MIME-Version: 1.0


Content-Type: text/plain
Content-Transfer-Encoding: 7bit

**Bold Text** Regular text

Content-Type: text/html
Content-Transfer-Encoding: 7bit

<p><strong>Bold Text</strong> Regular text</p>


"
`;

exports[`sendWelcomeEmail should match its snapshot 1`] = `
"Content-Type: multipart/alternative;

To: John Doe <test@example.com>
Subject: Welcome to Appsemble


MIME-Version: 1.0


Content-Type: text/plain
Content-Transfer-Encoding: 7bit

Hello John Doe,

Thank you for registering your account.  
Before you can use your account, we need to verify your email address.

Please follow the link below to verify your email address:  
https://example.com/test

Kind regards,

_The Appsemble Team_


Content-Type: text/html
Content-Transfer-Encoding: quoted-printable

<p>Hello John Doe,</p>
<p>Thank you for registering your account.<br>Before=
 you can use your account, we need to verify your email address.</p>
<p>Please follow the link below to verify your email address:<br><a =
href=3D\\"https://example.com/test\\">https://example.com/test</a></p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>


"
`;
